jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannelState.java

```diff
@@ -19,140 +19,223 @@
+
+            if (frame.getOpCode() == OpCode.CLOSE)
             {
-                State closed = new State("CLOSED", false, false, closeStatus);
-                if (state.compareAndSet(s, closed))
-                    return true;
+                _closeStatus = CloseStatus.getCloseStatus(frame);
+                _channelState = updateChannelState(_channelState, _incomingSequence, _outgoingSequence);
             }
+
+            return _channelState.equals(State.CLOSED);
         }
     }
 
-    public boolean onCloseOut(CloseStatus closeStatus)
+
+
+    public boolean checkIncoming(Frame frame) throws ProtocolException
     {
-        while (true)
+        synchronized (this)
         {
-            State s = state.get();
-
-            if (!s.outOpen)
-                throw new IllegalStateException(state.get().toString());
+            _incomingSequence = getNextState(frame, _incomingSequence);
 
-            if (s.inOpen)
-            {
-                State closedOut = new State("OCLOSED", true, false, closeStatus);
-                if (state.compareAndSet(s, closedOut))
-                    return false;
-            }
-            else
+            if (frame.getOpCode() == OpCode.CLOSE)
             {
-                State closed = new State("CLOSED", false, false, closeStatus);
-                if (state.compareAndSet(s, closed))
-                    return true;
+                _closeStatus = CloseStatus.getCloseStatus(frame);
+                _channelState = updateChannelState(_channelState, _incomingSequence, _outgoingSequence);
             }
+
+            return _channelState.equals(State.CLOSED);
         }
     }
 
-    public boolean onClosed(CloseStatus closeStatus)
+
+
+    private static State updateChannelState(State state, byte _incomingSequence, byte _outgoingSequence)
     {
-        while (true)
+        switch (state)
         {
-            State s = state.get();
-            if (!s.outOpen && !s.inOpen)
-                return false;
+            case OPEN:
+            case ICLOSED:
+            case OCLOSED:
+                break;
+            default:
+                throw new IllegalStateException(state.toString());
+        }
 
-            State newState = new State("CLOSED", false, false, closeStatus);
-            if (state.compareAndSet(s, newState))
-                return true;
+        State newState = state;
+        if ((_outgoingSequence == OpCode.CLOSE) && (_incomingSequence == OpCode.CLOSE))
+            newState = State.CLOSED;
+        else if (_outgoingSequence == OpCode.CLOSE)
+            newState = State.OCLOSED;
+        else if (_incomingSequence == OpCode.CLOSE)
+            newState = State.ICLOSED;
+
+        return newState;
+    }
+
+
+    public static byte getNextState(Frame frame, byte state) throws ProtocolException
+    {
+        byte opcode = frame.getOpCode();
+        boolean fin = frame.isFin();
+
+        if (state == OpCode.CLOSE)
```

Comment:
Doesn't this method need to know the real state if it is incoming or outgoing?  Receiving a frame in ICLOSED is ISE and sending a frame in OCLOSED is an ISE.   

Also recording closed state in both the `ICLOSED`, `OCLOSED` and `CLOSED` enums, plus in the two sequences is doubling up on state!   We should maintain one or the other, but not both.   Perhaps the sequence should only be used to record the type of continuations frames and not be used for close processing?